### DISCUSSION TITLE
Drawing with a device not assigned to the xinput core pointer

### DISCUSSION THREAD

**Discussion by coding-me-softly** [xournalpp/xournalpp#2810] [Q&A] [answered]
Hi, I am wondering how I might be able to draw with a device that is not a core pointer.  I usually like to `xinput float <device>` when I use my Wacom to stop it from being a secondary mouse.  Then, for example, I can specifically enable it again from within the GIMP so that it only works in that window.

My Wacom works fine in Xournalpp as long as it is a core pointer, but if not I haven't been able to get it to draw.  I've been tinkering with the Preferences (where the device is still recognized), but can't seem to get it to work.  Any suggestions for how I might get a non-core pointer device to work (if possible)?  I'm on 1.0.18 currently.

**Comment by coding-me-softly** [comment]
I've tried two additional things with `xinput` so far.

I tried to create a non-core pointer assigned to the Xournal window.  Maybe this works in other environments, but for me, doing that crashes my window manager and sends me back to the LightDM greeter.
```shell
$ xinput create-master wacom 0 1
$ xinput reattach 13 "wacom pointer" # the pen device
$ # at this point, moving the pointer with the pen crashes my WM, so I can't test further
```
I also created a core pointer assigned to the Xournal window.  Seems to work okay, but doesn't actually limit the pointer to just Xournal.  Also, drawing with both pointers at once makes some interesting geometric art.
```shell
$ xinput create-master wacom 1 1
$ xinput reattach 13 "wacom pointer"
$ xinput set-cp "Unsaved Document - Xournal++" "wacom pointer"
```

**Reply by Technius** [reply]
Are you using the new input system? @LittleHuba might be able to offer some better help.

**Reply by LittleHuba** [reply, marked as answer]
If you are using the new input system this is my point of view. I definitely advise you to switch to the nightly builds and try out these with the activated new input system. This might fix your issues.

That is something we can't influence very heavily I'm afraid. As far as I understand the WMs around, they rely on having a core pointer which is the master of all other pointer devices attached to the system. What we get from GTK+ is the core pointer of the system which allows us to enumerate all the slave pointers. From there on we treat all devices the same except how they behave within Xournal++.

> Also, drawing with both pointers at once makes some interesting geometric art.

Especially this makes me think that what you do is not exactly supported by GTK+ or that we use the devices in some specific way not working with your setup and therefore get very awkward results. All in all this is hard to fix and I'm not really sure if it is on us to fix this. This sounds more like something upstream.

**Comment by coding-me-softly** [comment]
Update for 2025: Applications shouldn't be using raw input events from slave devices anyway.  The most realistic way to accomplish what I described is to make a second core pointer and set up XFixes pointer barriers around the target area that change in response to ConfigureNotify events.  (No telling how laggy that would be either.)  Unfortunately XInput 2 device grabs don't support the confine-to parameter that is part of the base protocol's pointer grab.
